[Feature:Submission] Warn user when trying to upload more than the max allowed files

marwanatef2 opened the pull request:
Fixes #4456 

Users were allowed to add any no. of files they want however, the maximum allowable no. for submission in php is restricted by default to 20.
Now users are not allowed to add more than 20 files per submission where if they tried they get alerted that they have exceeded the limit.

The new behavior can be seen in this gif
![addvalid](https://user-images.githubusercontent.com/30902406/77235821-d71bb980-6bc1-11ea-9db4-85207f737e10.gif)


Please let me know if any messages or something need to be changed.

marwanatef2 commented:
> the best place to actually check for the file limit is the `handleSubmission` & `handlebulk` functions

my point is i don't want user to submit 20+ files then alert some message that submission file limit is exceeded so he has to remove some files, i want to prevent him from adding more than 20 files, stopping him at the limit, so it is better handled with addition not submission.

MasterOdin commented:
I agree on showing the message on adding, it's a friendlier user experience.

bmcutler commented:
Instead of hardcoding "20" into our source code, can we read the current system setting?
For example, if a production installation has modified the php.ini setting to be 50 files
https://www.codexworld.com/how-to/increase-files-upload-limit-in-php/

Also, can we add a line to the warning/error message that says
"Maximum allowed number of files to be uploaded is 20.
Please package your files in a single .zip file if it is necessary for you to submit more than this limit."

shailpatels commented:
@marwanatef2 are you still working on this PR?

marwanatef2 commented:
> @marwanatef2 are you still working on this PR?

Yes, I am. I'll get it done I was just working on my proposal.

marwanatef2 commented:
> Instead of hardcoding "20" into our source code, can we read the current system setting?

this part is done.

> Also, can we add a line to the warning/error message that says
> "Maximum allowed number of files to be uploaded is 20.
> Please package your files in a single .zip file if it is necessary for you to submit more than this limit."

this part can be shown in this screenshot
![Capture](https://user-images.githubusercontent.com/30902406/77843694-fa390100-719f-11ea-8b3d-b2f0ccc35c3f.PNG)

MasterOdin commented on site/app/templates/submission/homework/SubmitBox.twig:
```diff
@@ -2,7 +2,10 @@
-    <div class='content gradeable_message'>{{ gradeable_message | markdown }}</div>
+    <div class='content gradeable_message'>
+        {{ gradeable_message | markdown }}
+        Maximum allowed number of files to be uploaded is {{ max_file_uploads }}.
```
```
        {% if part_names|length > 1 %}
            <div id="upload-boxes">
                {# Submit boxes #}
                {% for part in part_names %}
                    {# uploadIndex and input-file + Index required for drag-and-drop.js #}
                    <div tabindex="0"
                        id="upload{{ loop.index }}"
                        class="upload-box"
                        onkeypress="clicked_on_box(event)"
                        role="text"
                        aria-label="Press enter to upload your {{ part }} file"
                    >
                        <h2 class="label" id="label{{ loop.index }}">
                            {% if part_names|length > 1 %}
                                Drag your {{ part }} file(s) here or click to open file browser
                            {% else %}
                                Drag your file(s) here or click to open file browser
                            {% endif %}
                        </h2>
                        <input type="file" name="files" id="input-file{{ loop.index }}" class="hide" onchange="addFilesFromInput({{ loop.index }})" multiple aria-label="Select Files to upload" />
                        <table class="file-upload-table" id="file-upload-table-{{ loop.index }}">
                        </table>
                    </div>
                {% endfor %}
                <p>Maximum allowed number of files to be uploaded is {{ max_file_uploads }}.</p>
            </div>
        {% endif %}
```

(I'm adding a surrounding `{% if %}` here as well as with your change.)

MasterOdin commented on site/public/js/drag-and-drop.js:
```diff
@@ -17,6 +17,9 @@ var label_array = [];
 var changed = false;        // if files from previous submission changed
 
+var total_files_added = 0;
+var MAX_NO_OF_FILES;
```
Use `MAX_NUM_OF_FILES` as the name.

Use `let` keyword please for new variables.

MasterOdin commented on site/public/js/drag-and-drop.js:
```diff
@@ -219,6 +246,7 @@ function deleteFiles(part) {
     var labels = dropzone.getElementsByClassName("file-label");
     while(labels[0]){
         dropzone.removeChild(labels[0]);
+        total_files_added --;
```
Get rid of the space here between the variable and the operator

MasterOdin commented on site/public/js/drag-and-drop.js:
```diff
@@ -243,6 +271,7 @@ function deleteSingleFile(filename, part, previous) {
             if (file_array[part-1][j].name == filename) {
                 file_array[part-1].splice(j, 1);
                 label_array[part-1].splice(j, 1);
+                total_files_added--;
```
Get rid of the space here between the variable and the operator

marwanatef2 commented:
> your message is causing the submission area to be pushed to the left side

it looks like I added the message outside the "upload box" div, modified now.
 
> On chrome after I submit past the max number of files, and get the alert, and I try to submit more than the max number again I don't get the alert, nothing seems to happen. This appears to be on chrome only, firefox does not have this issue on Mac.

I am working on chrome already and I had the same problem, It's fixed now as shown in the gif below.

![readding](https://user-images.githubusercontent.com/30902406/82282062-3d8b4100-9993-11ea-9a45-25df98380b41.gif)

bmcutler commented:
@shailpatels @MasterOdin 
Can one or both of you please re-review & approve
